Gems/Atom/RHI/Metal/Code/Source/RHI/CommandList.cpp

```diff
@@ -447,57 +400,129 @@ namespace AZ
 
-            id<MTLRenderCommandEncoder> renderEncoder = GetEncoder<id<MTLRenderCommandEncoder>>();
-            id<MTLComputeCommandEncoder> computeEncoder = GetEncoder<id<MTLComputeCommandEncoder>>();
-
-            //Call UseResource on all resources for Compute stage
-            for (const auto& key : resourcesToMakeResidentCompute)
+            if(bindNullDescriptorHeap)
             {
-                AZStd::vector<id <MTLResource>> resourcesToProcessVec(key.second.begin(), key.second.end());
-
-                [computeEncoder useResources: &resourcesToProcessVec[0]
-                                       count: resourcesToProcessVec.size()
-                                       usage: key.first];
-
+                MakeHeapsResident(mtlRenderStagesForNullDescHeap);
             }
+            return true;
+        }
 
-            //Call UseResource on all resources for Vertex and Fragment stages
-            for (const auto& key : resourcesToMakeResidentGraphics)
+        AZStd::pair<bool, id<MTLResource>> CommandList::GetResourceInfo(RHI::ShaderResourceGroupData::BindlessResourceType resourceType,
+                                                                        const RHI::ResourceView* resourceView)
+        {
+            id<MTLResource> mtlResourceView = nil;
+            bool isReadOnlyResource = false;
+            switch(resourceType)
             {
-
-                AZStd::vector<id <MTLResource>> resourcesToProcessVec(key.second.begin(), key.second.end());
-
-                [renderEncoder useResources: &resourcesToProcessVec[0]
-                                      count: resourcesToProcessVec.size()
-                                      usage: key.first.first
-                                     stages: key.first.second];
+                case RHI::ShaderResourceGroupData::BindlessResourceType::ReadTexture:
+                {
+                    isReadOnlyResource = true;
+                }
+                case RHI::ShaderResourceGroupData::BindlessResourceType::ReadWriteTexture:
+                {
+                    const ImageView* imageView = static_cast<const ImageView*>(resourceView);
+                    mtlResourceView = imageView->GetMemoryView().GetGpuAddress<id<MTLResource>>();
+                    break;
+                }
+                case RHI::ShaderResourceGroupData::BindlessResourceType::ReadBuffer:
+                {
+                    isReadOnlyResource = true;
+                }
+                case RHI::ShaderResourceGroupData::BindlessResourceType::ReadWriteBuffer:
+                {
+                    const BufferView* bufferView = static_cast<const BufferView*>(resourceView);
+                    mtlResourceView = bufferView->GetMemoryView().GetGpuAddress<id<MTLResource>>();
+                    break;
+                }
             }
-
-            if(bindNullDescriptorHeap)
+            return AZStd::pair<bool, id<MTLResource>>(isReadOnlyResource, mtlResourceView);
+        }
```

Comment:
why do read only resources return a nil resource view?

Reply:
They dont. There is no break so it falls into the next case statement.